sharktank/sharktank/utils/testing.py

```diff
@@ -108,6 +110,234 @@ def assertFileWritten(self, p: Path):
+
+        # NOTE: The _0 is required to match load_llm.py::Batch's syntax.
+        self.decode_iree_results_path = [work_dir / "decode_iree_logits.npy"]
+        self.decode_eager_logits_path = work_dir / "decode_eager_logits_0.npy"
+        self.decode_next_tokens_path = work_dir / "decode_next_tokens_0.npy"
+        self.decode_seq_lens_path = work_dir / "decode_seq_lens_0.npy"
+        self.decode_seq_block_ids_path = work_dir / "decode_seq_block_ids_0.npy"
+        self.decode_start_positions_path = work_dir / "decode_start_positions_0.npy"
+
+        self.prefill_cache_state_paths = []
+        self.decode_cache_state_paths = []
+        for i in range(
+            self.config.pipeline_parallelism_size * self.config.tensor_parallelism_size
+        ):
+            piece = f"_{i}" if i > 0 else ""
+            prefill_name = f"prefill_cache_state{piece}.npy"
+            decode_name = f"decode_cache_state_0{piece}.npy"  # _0 is required
+            self.prefill_cache_state_paths.append(work_dir / prefill_name)
+            self.decode_cache_state_paths.append(work_dir / decode_name)
+
+        self.dataset_path = work_dir / "parameters.irpa"
+
+        Dataset(root_theta=theta, properties=self.config.to_properties()).save(
+            path=self.dataset_path
+        )
+        self.exporter = ExportArtifacts.from_config(
+            self.config,
+            irpa_path=self.dataset_path,
+            iree_hip_target=self.iree_hip_target,
+            iree_hal_target_device=self.iree_hal_target_device,
+            hip_device_id=self.iree_device,
+            output_name=work_dir / "model",
+        )
+
+        self.config.device = torch.device("cuda:0")
+        self.inplace_transfer_theta(theta, self.config.device)
+
+        token_ids_prefill, seq_lens = pad_tokens(
+            token_ids=raw_token_ids,
+            pad_to_multiple_of=self.config.block_seq_stride,
+        )
+        token_ids_prefill = torch.as_tensor(
+            token_ids_prefill, device=self.config.device
+        )
+        seq_lens = torch.as_tensor(seq_lens, device=self.config.device)
+        self.batch_size = token_ids_prefill.shape[0]
+
+        generator = TorchGenerator(PagedLlmModelV1(theta=theta, config=self.config))
+        self.eager_batch = generator.begin_batch(
+            token_ids=token_ids_prefill, seq_lens=seq_lens, dump_path=work_dir
+        )
+
+    def run_eager(self):
+        """
+        Run the eager execution of the LLM prefill and decode stages.
+        Saveing the cache state before each stage to be used in IREE execution.
+        """
+        eager_decode_tokens = self.eager_batch.prefill()
+        self.eager_batch.dump_args(
+            phase="prefill",
+            arg_name="eager_logits",
+            arg=self.eager_batch.prefill_logits,
+        )
+
+        self.eager_prefill_logits = torch.tensor(
+            np.load(self.prefill_eager_logits_path)
```

Comment:
bfloat16 and the various f8 types would be a problem form numpy.
You could use [safetensors](https://github.com/huggingface/safetensors?tab=readme-ov-file#yet-another-format-).

Reply:
I am not sure if you could pass safetensors as iree-run-module function args. I recall you can't. Maybe it got support.
If not, you could use [iree_tool_prepare_input_args](https://github.com/iree-org/iree-turbine/blob/9cc8c9a53d910504581e26170565a63489ff12df/iree/turbine/support/tools.py#L36C5-L36C33), which would support any torch dtype.

Reply:
I am relying on `load_llm.py:Batch` to run the torch model and dump these values. If fp8 support requires special work then I think that should be in a subsequent PR as it will require redoing portions of the `Batch` class (at a minimum).